File: CustomApplications/CSharp/GraphicsHowTo/DataPath.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;
using System.Runtime.InteropServices;
using System.Configuration;
using System.Windows.Forms;

namespace AGI
{
    public enum DataPathRoot
    {
        TestData,
        Relative
    }

    [Serializable]
    public class DataPath
    {
        public DataPath(DataPathRoot root, string relative)
        {
            m_root = root;
            m_relative = relative;
        }

        public string FullPath
        {
            get
            {
                string root;
                if (m_root == DataPathRoot.TestData)
                {
                    // all users
                    string allUsers = System.Environment.GetEnvironmentVariable("ALLUSERSPROFILE");
                    // rest of path added a few lines below
                    string dataDir = allUsers;
                    bool notVista = (System.Environment.GetEnvironmentVariable("ProgramData") == null);
                    if (notVista)
                    {
                        dataDir = allUsers + @"\Application Data";
                    }
                    root = dataDir + @"\AGI\STK 11\";
                }
                else
                {
                    // Set a path to a directory containing the test resources.
                    root = Path.Combine(Application.StartupPath, "Data");
                }

                string relative = m_relative.Replace('\\', Path.DirectorySeparatorChar);
                relative = relative.Replace('/', Path.DirectorySeparatorChar);

                return Path.Combine(root, relative);
            }
        }

        public Uri Uri
        {
            get { return new Uri(FullAbsolutePath); }
        }

        public string FullAbsolutePath
        {
            get
            {
                string fullPath = FullPath;
                if (string.IsNullOrEmpty(fullPath))
                    return Directory.GetCurrentDirectory() + Path.DirectorySeparatorChar;

                if (!Path.IsPathRooted(fullPath))
                    return Path.Combine(Directory.GetCurrentDirectory(), fullPath);

                return fullPath;
            }
        }

        public override bool Equals(object obj)
        {
            DataPath dp = obj as DataPath;
            if (dp != null)
            {
                return FullPath == dp.FullPath;
            }
            else
            {
                return false;
            }
        }

        public override int GetHashCode()
        {
            return m_root.GetHashCode() ^ m_relative.GetHashCode();
        }

        private DataPathRoot m_root;
        private string m_relative;
    }
}

File: CustomApplications/VB.Net/GraphicsHowTo/DataPath.vb
Imports System.Collections.Generic
Imports System.Text
Imports System.IO
Imports System.Runtime.InteropServices
Imports System.Configuration
Imports System.Windows.Forms
Imports AGI.STKObjects
Imports AGI.STKUtil

Namespace AGI
	Public Enum DataPathRoot
		TestData
		Relative
	End Enum

	<Serializable> _
	Public Class DataPath
		Public Sub New(root As DataPathRoot, relative As String)
			m_root = root
			m_relative = relative
		End Sub

		Public ReadOnly Property FullPath() As String
			Get
                Dim root As String
                If m_root = DataPathRoot.TestData Then
                    ' all users
                    Dim allUsers As String = System.Environment.GetEnvironmentVariable("ALLUSERSPROFILE")
                    ' rest of path added a few lines below
                    Dim dataDir As String = allUsers
                    Dim notVista As Boolean = (System.Environment.GetEnvironmentVariable("ProgramData") Is Nothing)
                    If notVista Then
                        dataDir = allUsers & "\Application Data"
                    End If
                    root = dataDir & "\AGI\STK 11\"
                Else
                    ' Set a path to a directory containing the test resources.
                    root = Path.Combine(Application.StartupPath, "Data")
                End If

                Dim relative As String = m_relative.Replace("\"c, Path.DirectorySeparatorChar)
                relative = relative.Replace("/"c, Path.DirectorySeparatorChar)

                Return Path.Combine(root, relative)
			End Get
		End Property

		Public ReadOnly Property Uri() As Uri
			Get
				Return New Uri(FullAbsolutePath)
			End Get
		End Property

		Public ReadOnly Property FullAbsolutePath() As String
			Get
				Dim fullPath__1 As String = FullPath
				If String.IsNullOrEmpty(fullPath__1) Then
					Return Directory.GetCurrentDirectory() & Path.DirectorySeparatorChar
				End If

				If Not Path.IsPathRooted(fullPath__1) Then
					Return Path.Combine(Directory.GetCurrentDirectory(), fullPath__1)
				End If

				Return fullPath__1
			End Get
		End Property

		Public Overrides Function Equals(obj As Object) As Boolean
			Dim dp As DataPath = TryCast(obj, DataPath)
			If dp IsNot Nothing Then
				Return FullPath = dp.FullPath
			Else
				Return False
			End If
		End Function

		Public Overrides Function GetHashCode() As Integer
			Return m_root.GetHashCode() Xor m_relative.GetHashCode()
		End Function

		Private m_root As DataPathRoot
		Private m_relative As String
	End Class
End Namespace
